lib/spack/spack/abi.py

```diff
@@ -21,7 +21,7 @@ class ABI(object):
     def architecture_compatible(self, parent, child):
         """Return true if parent and child have ABI compatible targets."""
         return not parent.architecture or not child.architecture or \
-            parent.architecture == child.architecture
+            parent.architecture.satisfies(child.architecture)
```

Comment:
This doesn't seem completely right, at least according to the current docstrings. What I mean is that saying that "A and B are compatible" means to me "A compatible with B" and "B compatible with A". Here we are removing the compatibility in one direction.

What I mean is that probably the change is functionally correct since this method is only indirectly called in:

https://github.com/spack/spack/blob/dd3378bda4dfb4a1ba0344ea50bf472436b41538/lib/spack/spack/concretize.py#L116-L137

but I think we need to make this clear in the docstring here and in the `compatible` method below and maybe change the methods name too.

Reply:
* Rebased on current `develop`
* Clarified the non-commutativity of `spack.abi.ABI.architecture_compatible` and `spack.abi.ABI.compatible` in the docstrings
* Added some standalone unit test coverage for `spack.abi`